src/utils/recompiler.ts

```diff
@@ -0,0 +1,127 @@
+        // create file, if it's not found
+        await fsPromises.writeFile(dirpath, JSON.stringify({}));
+    } else {
+        // try to read file
+        const oldFile = await fsPromises.readFile(dirpath);
+        oldNameHashPair = JSON.parse(oldFile.toString() || "{}");
+    }
+};
+
+// Gets hash of each .cairo file inside contracts
+const getContractHash = async (defaultSourcesPath: string) => {
+    try {
+        // traverse directory contracts/
+        const files = await fsPromises.readdir(defaultSourcesPath);
+        // check only cairo file extensions
+        const filesList = files.filter((el) => path.extname(el).toLowerCase() === ".cairo");
+        // select file name
+        for (const cairoContract of filesList) {
+            const data = await fsPromises.readFile(defaultSourcesPath.concat(`/${cairoContract}`));
+            const hash = createHash("sha256");
+            hash.update(data);
+            newNameHashPair[cairoContract] = hash.digest("hex").toString();
+        }
+    } catch (err) {
+        console.log(err);
+    }
+};
+
+// Checks artifacts availability
+const checkArtifacts = async (defaultSourcesPath: string, starknetArtifacts: string) => {
+    try {
+        // traverse on artifacts directory
+        const dirpath = path.join(starknetArtifacts, "contracts");
+        const files = await fsPromises.readdir(dirpath);
+        for (const name in newNameHashPair) {
+            if (!files.includes(name)) {
+                changed.add(`${defaultSourcesPath}/${name}`);
+            }
+        }
+    } catch (err) {
+        console.log(err);
+    }
+};
+
+// Compile changed contracts
+const compileChangedContracts = async (args: TaskArguments, hre: HardhatRuntimeEnvironment): Promise<boolean> => {
+    const defaultSourcesPath = hre.config.paths.starknetSources;
+    const getKeys = Object.keys;
+
+    getKeys(newNameHashPair).forEach(contractName => {
+        // Add new contracts that are not in cache before
+        if (!oldNameHashPair[contractName]) {
+            changed.add(`${defaultSourcesPath}/${contractName}`);
+        }
+
+        // Add contracts that contiain a change in content
+        if (newNameHashPair[contractName] !== oldNameHashPair[contractName]) {
+            changed.add(`${defaultSourcesPath}/${contractName}`);
+        }
+    });
+
+    if (changed.size > 0) {
+        try {
+            // Compiles contracts
+            console.log("Compiling contracts...");
+            args.paths = [...changed];
+            await starknetCompileAction(args, hre);
+        } catch (err) {
+            console.log(err);
+            return false;
+        }
+    }
+    return true;
+};
+
+export const handleCache = async (args: TaskArguments, hre: HardhatRuntimeEnvironment): Promise<boolean> => {
+    const { starknetSources: defaultSourcesPath, cache: cacheDirName, starknetArtifacts } = hre.config.paths;
+
+    await upsertFile(cacheDirName);
+    await getContractHash(defaultSourcesPath);
```

Comment:
No need for these three functions to communicate over global variables. Their return values could be used for exchanging data structures.